src/ds3.c

```diff
@@ -109,24 +111,27 @@ static void _ds3_free_metadata_entry(gpointer pointer) {
 }
 
 /*
- * This copies all the header values in the ds3_response_header struct so that they may be safely returned to the user
+ * This copies all the header values in the ds3_string_multimap_entry struct so that they may be safely returned to the user
  * without having to worry about if the data is freed internally.
  */
-static ds3_metadata_entry* ds3_metadata_entry_init(ds3_response_header* header) {
+static ds3_metadata_entry* ds3_metadata_entry_init(ds3_string_multimap_entry* header_entry) {
     guint i;
     ds3_str* header_value;
     GPtrArray* values = g_ptr_array_new();
     ds3_str* key_name;
     ds3_metadata_entry* response = g_new0(ds3_metadata_entry, 1);
 
-    for (i = 0; i < header->values->len; i++) {
-        header_value = (ds3_str*)g_ptr_array_index(header->values, i);
-        g_ptr_array_add(values, ds3_str_dup(header_value));
+    unsigned int num_values = ds3_string_multimap_entry_get_num_values(header_entry);
+    for (i = 0; i < num_values; i++) {
+        header_value = ds3_string_multimap_entry_get_value_by_index(header_entry, i);
+        g_ptr_array_add(values, header_value);
     }
 
-    key_name = ds3_str_init(header->key->value + 11);
+    ds3_str* full_key = ds3_string_multimap_entry_get_key(header_entry);
+    key_name = ds3_str_init(full_key->value + 11); // offset after "x-amz-meta-"
```

Comment:
I noticed that the string "x-amz-meta-" appears bellow.  Should it be placed in a static string at the beginning of the file, and referenced here by length instead of using the magic number 11?